lib/application/widgets/theme/custom_button.dart

```diff
@@ -0,0 +1,313 @@
+  Widget build(BuildContext context) {
+    final theme = useTheme();
+    final buttonColorSwatch = color ?? theme.primarySwatch;
+    final textTheme = Theme.of(context).textTheme.button!;
+
+    Color? buttonColor(ButtonState state) {
+      switch (state) {
+        case ButtonState.normal:
+          return buttonColorSwatch.shade300;
+        case ButtonState.pressed:
+          return buttonColorSwatch.shade400;
+        case ButtonState.disabled:
+          return buttonColorSwatch.shade400.withOpacity(0.4);
+      }
+    }
+
+    Widget leadingAssetBuilder(ButtonState state) => Image.asset(leadingAsset!, color: buttonColor(state));
+
+    return CustomButton(
+      text: text,
+      leadingWidgetBuilder: leadingAsset != null ? leadingAssetBuilder : null,
+      onPressed: onPressed,
+      textStyleBuilder: (state) => textTheme.copyWith(color: buttonColor(state)),
+    );
+  }
+}
+
+/// Wraps a [CustomButton] following a similar style to [ElevatedButton].
+///
+/// Reused by custom buttons implementations.
+/// See also:
+///
+///   * [PrimaryElevatedButton]
+///   * [SecondaryElevatedButton]
+class _CustomElevatedButton extends StatelessWidget {
+  const _CustomElevatedButton({
+    required this.text,
+    required this.backgroundColorBuilder,
+    this.backgroundColor,
+    this.onPressed,
+    this.leadingAsset,
+  });
+
+  final String text;
+
+  /// {@template _CustomElevatedButton.backgroundColor}
+  /// Button's background color swatch.
+  /// {@endtemplate}
+  final MaterialColor? backgroundColor;
+  final VoidCallback? onPressed;
+
+  /// {@template _CustomElevatedButton.leadingAsset}
+  /// Optional leading asset.
+  ///
+  /// Its color will be overriden by the button `TextStyle`.
+  /// {@endtemplate}
+  final String? leadingAsset;
+
+  /// Builds the button background color based on the button state.
+  final Color? Function(ButtonState state) backgroundColorBuilder;
+
+  @override
+  Widget build(BuildContext context) {
+    final textTheme = Theme.of(context).textTheme.button!;
+
+    Widget leadingAssetBuilder(ButtonState state) => Image.asset(leadingAsset!, color: textTheme.color);
+
+    return CustomButton(
+      text: text,
+      onPressed: onPressed,
+      isPressedOverlayEnabled: true,
+      backgroundColorBuilder: backgroundColorBuilder,
+      textStyleBuilder: (state) => textTheme,
+      leadingWidgetBuilder: leadingAsset != null ? leadingAssetBuilder : null,
+    );
+  }
+}
+
+/// Shared visual style between all custom button's.
+///
+/// Provides a common and customizable button style that must be used by all custom buttons.
+///
+/// See also:
+///
+///   * [PrimaryElevatedButton] - The primary visually opinionated alternative to [ElevatedButton].
+///   * [SecondaryElevatedButton] - The secondary visually opinionated alternative to [ElevatedButton].
+///   * [CustomTextButton] - The visually opinionated alternative to [TextButton].
+@visibleForTesting
+class CustomButton extends StatefulWidget {
```

Comment:
I think keeping this private and testing all implementations (primary, secondary and text) directly would be better here, as the usages of this "internal" behavior is what matters for us. You could test both as well, but I don't know if it would make sense

Reply:
To do that we would have to change the testing strategy to detect the state change based on the Widget interface update, because making `CustomButton` makes us loose access to its `state` property. The problem that I see here is that if we decide to change the buttons interface it would directly affect the tests, while testing by `CustomButton.state` test the behavior itself, independent from the UI

Reply:
And that's fine, because unit tests are affordable to be deleted/updated/added, but we want to make sure that the functionality itself is working as intended, and not some particular reusable internal state, but I do see where you're getting at, and, if necessary, I think it could be solved by structuring the tests where some of its logic could be shared/reused by all of these interfaces